torch/testing/_internal/common_modules.py

```diff
@@ -0,0 +1,203 @@
+                return reference_fn(m, list(m.parameters()), *args, **kwargs)
+
+            self.reference_fn = copy_reference_fn
+
+
+class ModuleInfo(object):
+    """ Module information to be used in testing. """
+
+    def __init__(self,
+                 module_cls,  # Class object for the module under test
+                 *,
+                 module_inputs_func,  # Function to generate module inputs
+                 skips=[],  # Indicates which tests to skip
+                 decorators=None,  # Additional decorators to apply to generated tests
+                 ):
+        self.module_cls = module_cls
+        self.module_inputs_func = module_inputs_func
+        self.skips = skips
+        self.decorators = decorators
+
+    def should_skip(self, cls_name, test_name, device_type, dtype):
+        return any(si.is_active(cls_name, test_name, device_type, dtype) for si in self.skips)
+
+    @property
+    def name(self):
+        return formatted_module_name(self.module_cls)
+
+
+def module_inputs_torch_nn_Linear(module_info, device, dtype, requires_grad, **kwargs):
+    make_input = partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad)
+
+    module_inputs = [
+        ModuleInput(constructor_input=FunctionInput(10, 8),
+                    forward_input=FunctionInput(make_input((4, 10))),
+                    reference_fn=lambda m, p, i: torch.mm(i, p[0].t()) + p[1].view(1, -1).expand(4, 8)),
+        ModuleInput(constructor_input=FunctionInput(10, 8, bias=False),
+                    forward_input=FunctionInput(make_input((4, 10))),
+                    desc='no_bias',
+                    reference_fn=lambda m, p, i: torch.mm(i, p[0].t())),
+        ModuleInput(constructor_input=FunctionInput(3, 5),
+                    forward_input=FunctionInput(make_input(3)),
+                    desc='no_batch_dim',
+                    reference_fn=lambda m, p, i: torch.mm(i.view(1, -1), p[0].t()).view(-1) + p[1])
+    ]
+
+    return module_inputs
+
+
+def module_inputs_torch_nn_NLLLoss(module_info, device, dtype, requires_grad, **kwargs):
+    cases = [
+        ('', {}),
+        ('ignore_index', {'ignore_index': 2}),
+        ('weights', {'weight': torch.randn(10, device=device, dtype=dtype)}),
+        ('weights_ignore_index', {'weight': torch.randn(10, device=device, dtype=dtype), 'ignore_index': 2}),
+        ('weights_ignore_index_neg', {'weight': torch.randn(10, device=device, dtype=dtype), 'ignore_index': -1})
+    ]
+    module_inputs = []
+    for desc, constructor_kwargs in cases:
+
+        def reference_fn(m, p, i, t, constructor_kwargs=constructor_kwargs):
+            return nllloss_reference(i, t, **constructor_kwargs)
+
+        module_inputs.append(
+            ModuleInput(constructor_input=FunctionInput(**constructor_kwargs),
+                        forward_input=FunctionInput(torch.rand(15, 10, device=device, dtype=dtype).log_softmax(dim=1),
```

Comment:
In the future you may want to consider using `make_tensor` instead of `rand`, but no reason to change now